src/Models/HydrostaticFreeSurfaceModels/split_explicit_free_surface.jl

```diff
@@ -132,55 +178,117 @@ function SplitExplicitAuxiliary(grid::AbstractGrid)
+    struct SplitExplicitSettings
 
 A struct containing settings for the split-explicit free surface.
 
 $(TYPEDFIELDS)
 """
-struct SplitExplicitSettings{𝒩, ℳ}
+struct SplitExplicitSettings{𝒩, ℳ, 𝒯, 𝒮}
     "substeps: (`Int`)"
     substeps :: 𝒩
-    "velocity_weights : (`Vector`)"
-    velocity_weights :: ℳ
-    "free_surface_weights : (`Vector`)"
-    free_surface_weights :: ℳ
+    "averaging_weights : (`Vector`)"
+    averaging_weights :: ℳ
+    "mass_flux_weights : (`Vector`)"
+    mass_flux_weights :: ℳ
+    "fractional step: (`Number`), the barotropic time step will be (Δτ ⋅ Δt)" 
+    Δτ :: 𝒯
+    "time stepping scheme"
+    timestepper :: 𝒮
 end
 
-function SplitExplicitSettings(; substeps = 200, velocity_weights = nothing, free_surface_weights = nothing)
-    velocity_weights = Tuple(ones(substeps) ./ substeps)
-    free_surface_weights = Tuple(ones(substeps) ./ substeps)
+"""
+    Possible barotropic time-stepping scheme. 
+
+- `AdamsBashforth3Scheme`: η = f(U, Uᵐ⁻¹, Uᵐ⁻²) then U = f(η, ηᵐ, ηᵐ⁻¹, ηᵐ⁻²)
+- `ForwardBackwardScheme`: η = f(U)             then U = f(η)
+"""
+struct AdamsBashforth3Scheme end
+struct ForwardBackwardScheme end
 
-    return SplitExplicitSettings(substeps,
-        velocity_weights,
-        free_surface_weights)
+# Weights that minimize dispersion error from http://falk.ucsd.edu/roms_class/shchepetkin04.pdf (p = 2, q = 4, r = 0.18927)
+@inline function averaging_shape_function(τ; p = 2, q = 4, r = 0.18927) 
+    τ₀ = (p + 2) * (p + q + 2) / (p + 1) / (p + q + 1) 
+    return (τ / τ₀)^p * (1 - (τ / τ₀)^q) - r * (τ / τ₀)
 end
 
-function SplitExplicitSettings(substeps)
-    velocity_weights = Tuple(ones(substeps) ./ substeps)
-    free_surface_weights = Tuple(ones(substeps) ./ substeps)
+@inline cosine_averaging_kernel(τ::FT) where FT = τ >= 0.5 && τ <= 1.5 ? FT(1 + cos(2π * (τ - 1))) : zero(FT)
+
+@inline constant_averaging_kernel(τ) = 1
 
-    return SplitExplicitSettings(substeps = substeps,
-        velocity_weights = velocity_weights,
-        free_surface_weights = free_surface_weights)
+"""
+    SplitExplicitSettings(; substeps = 200, 
+                            barotropic_averaging_kernel = averaging_shape_function,
+                            timestepper = ForwardBackwardScheme())
+
+constructor for `SplitExplicitSettings`. For a description of the keyword
+arguments, see the `SplitExplicitFreeSurface` constructor
+"""
+function SplitExplicitSettings(; substeps = 200, 
+                                 barotropic_averaging_kernel = averaging_shape_function,
+                                 timestepper = ForwardBackwardScheme())
+
+    τᶠ = range(0, 2, length = substeps+1)
+    Δτ = τᶠ[2] - τᶠ[1]
+
+    averaging_weights = barotropic_averaging_kernel.(τᶠ[2:end]) 
+    idx = searchsortedlast(averaging_weights, 0.0, rev=true)
+    substeps = idx
+
+    averaging_weights = averaging_weights[1:idx]
+    mass_flux_weights = similar(averaging_weights)
+    
+    M = searchsortedfirst(τᶠ, 1.0) - 1
```

Comment:
```suggestion
    M = searchsortedfirst(τᶠ, 1) - 1
```

Reply:
These changes are just to make it easier to diagnose type instability bugs in the future